Library/ViewModels/CommentsViewModelTests.swift

```diff
@@ -111,365 +113,404 @@ internal final class CommentsViewModelTests: TestCase {
+      self.isCommentsLoading.assertValues([true, false], "loading ends")
+    }
+  }
 
-        self.loadCommentsAndProjectIntoDataSourceComments.assertDidNotEmitValue()
-        self.loadCommentsAndProjectIntoDataSourceProject.assertDidNotEmitValue()
-        self.isCommentsLoading.assertValues([true], "loading begins")
+  func testComments_WhenNoProjectOrUpdate_CommentsAreNotUpdatedInDataSource() {
+    self.loadCommentsAndProjectIntoDataSourceComments.assertDidNotEmitValue()
+    self.loadCommentsAndProjectIntoDataSourceProject.assertDidNotEmitValue()
 
-        self.scheduler.advance()
+    let envelope = CommentsEnvelope.singleCommentTemplate
 
-        self.loadCommentsAndProjectIntoDataSourceComments
-          .assertValues([envelope.comments], "New comments are emitted")
-        self.loadCommentsAndProjectIntoDataSourceProject
-          .assertValues([.template], "Same project is emitted again")
-        self.isCommentsLoading.assertValues([true, false], "loading ends")
-      }
-    }
+    withEnvironment(apiService: MockService(fetchCommentsEnvelopeResult: .success(envelope))) {
+      self.vm.inputs.configureWith(
+        project: nil,
+        update: nil
+      )
+      self.vm.inputs.viewDidLoad()
 
-    func testComments_WhenNoProjectOrUpdate_CommentsAreNotUpdatedInDataSource() {
       self.loadCommentsAndProjectIntoDataSourceComments.assertDidNotEmitValue()
       self.loadCommentsAndProjectIntoDataSourceProject.assertDidNotEmitValue()
+      self.isCommentsLoading.assertDidNotEmitValue()
 
-      let envelope = CommentsEnvelope.singleCommentTemplate
+      self.scheduler.advance()
 
-      withEnvironment(apiService: MockService(fetchCommentsEnvelopeResult: .success(envelope))) {
-        self.vm.inputs.configureWith(
-          project: nil,
-          update: nil
-        )
-        self.vm.inputs.viewDidLoad()
+      self.loadCommentsAndProjectIntoDataSourceComments.assertDidNotEmitValue()
+      self.loadCommentsAndProjectIntoDataSourceProject.assertDidNotEmitValue()
+      self.isCommentsLoading.assertDidNotEmitValue()
+    }
+  }
 
-        self.loadCommentsAndProjectIntoDataSourceComments.assertDidNotEmitValue()
-        self.loadCommentsAndProjectIntoDataSourceProject.assertDidNotEmitValue()
-        self.isCommentsLoading.assertDidNotEmitValue()
+  func testOptimisticComment() {
+    self.vm.inputs.configureWith(project: .template, update: nil)
 
-        self.scheduler.advance()
+    let envelope = CommentsEnvelope.singleCommentTemplate
 
-        self.loadCommentsAndProjectIntoDataSourceComments.assertDidNotEmitValue()
-        self.loadCommentsAndProjectIntoDataSourceProject.assertDidNotEmitValue()
-        self.isCommentsLoading.assertDidNotEmitValue()
-      }
+    withEnvironment(
+      apiService: MockService(fetchCommentsEnvelopeResult: .success(envelope)),
+      currentUser: .template
+    ) {
+      self.vm.inputs.viewDidLoad()
+
+      self.scheduler.advance()
```

Comment:
Before `advancing` test that the optimistic comment is not returned by `loadCommentsAndProjectIntoDataSourceComments`